plugins/modeltraining/pkg/modeltraining/aggregation.go

```diff
@@ -0,0 +1,169 @@
+package modeltraining
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"log"
+	"time"
+)
+
+type Key struct {
+	ClusterID      string `json:"cluster_id"`
+	NamespaceName  string `json:"namespace_name"`
+	DeploymentName string `json:"deployment_name"`
+}
+
+type SearchResponse struct {
+	Aggregations AggregationsSpec `json:"aggregations"`
+}
+
+type AggregationsSpec struct {
+	Bucket BucketSpec `json:"bucket"`
+}
+
+type BucketSpec struct {
+	AfterKey *Key     `json:"after_key"`
+	Buckets  []Bucket `json:"buckets"`
+}
+
+type Bucket struct {
+	Key      Key `json:"key"`
+	DocCount int `json:"doc_count"`
+}
+
+type Aggregations struct {
+	ByCluster map[string]*ClusterAggregation `json:",inline,omitEmpty"`
+}
+
+type ClusterAggregation struct {
+	ByNamespace map[string]*NamespaceAggregation `json:",inline,omitEmpty"`
+}
+
+type NamespaceAggregation struct {
+	ByDeployment map[string]*DeploymentLogCount `json:",inline,omitEmpty"`
+}
+
+type DeploymentLogCount struct {
+	DeploymentName string `json:"deployment_name"`
+	Count          int    `json:"doc_count"`
+}
+
+func newAggregations() *Aggregations {
+	return &Aggregations{ByCluster: make(map[string]*ClusterAggregation)}
+}
+
+func (a *Aggregations) Add(bucket Bucket) {
+	if _, ok := a.ByCluster[bucket.Key.ClusterID]; !ok {
+		a.ByCluster[bucket.Key.ClusterID] = &ClusterAggregation{
+			ByNamespace: make(map[string]*NamespaceAggregation),
+		}
+	}
+	cluster := a.ByCluster[bucket.Key.ClusterID]
+	if _, ok := cluster.ByNamespace[bucket.Key.NamespaceName]; !ok {
+		cluster.ByNamespace[bucket.Key.NamespaceName] = &NamespaceAggregation{
+			ByDeployment: make(map[string]*DeploymentLogCount),
+		}
+	}
+	a.ByCluster[bucket.Key.ClusterID].ByNamespace[bucket.Key.NamespaceName].ByDeployment[bucket.Key.DeploymentName] = &DeploymentLogCount{
+		DeploymentName: bucket.Key.DeploymentName,
+		Count:          bucket.DocCount,
+	}
+}
+
+func (s *ModelTrainingPlugin) run_aggregation() {
+	for {
+		request := map[string]interface{}{
```

Comment:
This should either be defined using a custom struct or a json string, nested `map[string]any` is kind of clunky.

Reply:
https://github.com/rancher/opni/issues/724